receiver/udp.go

```diff
@@ -121,3 +153,45 @@ func (ud *UDP) Start() error {
 		ud.ch <- bytes[0:n]
 	}
 }
+
+// GetStats prepares a skogul metric with stats
+// for the UDP receiver.
+func (ud *UDP) GetStats() *skogul.Metric {
+	metric := ud.stats.Generic.Metric()
+	metric.Metadata["component"] = "receiver"
+	metric.Metadata["type"] = "UDP"
+	metric.Metadata["name"] = "N/A"
+	return metric
+}
+
+// initStats initializes up the necessary components for stats
+func (ud *UDP) initStats() {
+	ud.stats = &udpStats{
+		Generic: &skogul.Stats{},
+	}
+	ud.stats.ch = make(chan stat, 10*ud.Threads)
+	ud.ticker = time.NewTicker(ud.EmitStats.Duration)
+}
+
+// startStats starts the listening loop for the stats chan
+func (ud *UDP) startStats() {
+	for {
+		switch <-ud.stats.ch {
+		case receive:
+			ud.stats.Generic.Received++
+		case er:
+			ud.stats.Generic.Errors++
+		case sent:
+			ud.stats.Generic.Sent++
```

Comment:
I don't really see the benefit of using a channel and a secondary thread here. Wouldn't atomic.AddUint64 for instance do just as well, but with lower overhead? I know channels are fast, but it's difficult to believe they are faster than an atomic +1.

Reply:
I haven't done any benchmarking, but it sounds possible that it has a lower overhead. I'm not heavily invested in either approach, so I'll swap it. Two of my "pros" from the channel approach was that it is possible to create an "enum" type which makes it very easy to read how stats are generated/incremented, compared to the more verbose atomic.AddUint64() (`ud.stats.ch <- received` vs. `atomic.AddUint64(ud.stats.Received, 1)`). The other is that there is a (symbolic?) separation of concerns between who counts the stats and who generates them, in different goroutines. But that might add more overhead. 🤷‍♂️